### PR TITLE
Zoom improvements

### PR DESCRIPTION
**Pull request by danielcaceresm**
Thanks you for this useful library. We did some changes to integrate it in our WaveEngine Editor, and we are working to create pull requests with them. This is the first one adding some features to DragCanvas.

- Added Min/Max and DragOffset dependency properties to DragCanvas.
- Added SetViewport method to DragCanvas.
- ShaderEditor example modified to add a toolbar showing these features.

### CODE REVIEW

File: NodeNetwork/ViewModels/NetworkViewModel.cs [hunk 1]
```diff
@@ -151,22 +151,49 @@ public bool IsReadOnly
 
         #region ZoomFactor
         /// <summary>
-        /// Scale of the view. Larger means more zoomed in. Default value is 1. Must be greater than zero.
+        /// Scale of the view. Larger means more zoomed in. Default value is 1.
         /// </summary>
-        public double ZoomFactor
+        public double? ZoomFactor
```
Comment (Wouterdek) [inline NodeNetwork/ViewModels/NetworkViewModel.cs:156]: It's not clear to me why this is nullable, or what happens if I set to null? Please either add some documentation on that behaviour, or make it non-nullable. Perhaps also explain that values outside the min/max range get clamped.

File: NodeNetwork/ViewModels/NetworkViewModel.cs [hunk 2]
```diff
@@ -151,22 +151,49 @@ public bool IsReadOnly
...
-                }
-                this.RaiseAndSetIfChanged(ref _zoomFactor, value);
-            }
+            set => this.RaiseAndSetIfChanged(ref _zoomFactor, value);
+        }
+
+        private double? _zoomFactor = 1;
+
+        /// <summary>
+        /// The maximun zoom level used in this network view. Default value is
```
Comment (Wouterdek) [inline NodeNetwork/ViewModels/NetworkViewModel.cs:165]: Nit. typo: maximun, and missing default value.

File: NodeNetwork/ViewModels/NetworkViewModel.cs [hunk 3]
```diff
@@ -151,22 +151,49 @@ public bool IsReadOnly
...
+        {
+            get => _maxZoomLevel;
+            set => this.RaiseAndSetIfChanged(ref _maxZoomLevel, value);
         }
 
-        private double _zoomFactor = 1;
+        private double? _maxZoomLevel;
+
+        /// <summary>
+        /// The minimun zoom level used in this network view.
```
Comment (Wouterdek) [inline NodeNetwork/ViewModels/NetworkViewModel.cs:176]: Same as above.

File: NodeNetwork/ViewModels/NetworkViewModel.cs [hunk 4]
```diff
@@ -151,22 +151,49 @@ public bool IsReadOnly
...
+            get => _minZoomLevel;
+            set => this.RaiseAndSetIfChanged(ref _minZoomLevel, value);
+        }
+
+        private double? _minZoomLevel;
+
+        /// <summary>
+        /// The position used in this network view.
+        /// </summary>
+        public Point? Position
```
Comment (Wouterdek) [inline NodeNetwork/ViewModels/NetworkViewModel.cs:189]: "Position" doesn't really explain the use of the variable to me here, and neither does the documentation. From reading the code I assume this refers to the offset between the origin of the canvas and the origin of the nodes, which is modified during dragging? A better name might then be "DragOffset" or something like that. Also, please add a bit more explanation to the documentation.

File: NodeNetwork/Views/Controls/DragCanvas.cs [hunk 5]
```diff
@@ -1,20 +1,36 @@
...
 
 namespace NodeNetwork.Views.Controls
 {
     public class DragCanvas : Canvas
     {
+        #region Position
+        public static readonly DependencyProperty PositionOffsetProperty = DependencyProperty.Register(nameof(PositionOffset),
+            typeof(Point), typeof(DragCanvas), new PropertyMetadata(new Point(), PositionOffsetChanged));
+
+        public Point PositionOffset
```
Comment (Wouterdek) [inline NodeNetwork/Views/Controls/DragCanvas.cs:16]: Same comment as with "Position" above. There is no documentation here, and the name doesn't really fully cover what this variable does.

File: NodeNetwork/Views/Controls/DragCanvas.cs [hunk 6]
```diff
@@ -102,7 +119,8 @@ protected override void OnMouseMove(MouseEventArgs e)
                     var dragEvent = new DragMoveEventArgs(e, xDelta, yDelta);
                     DragMove?.Invoke(this, dragEvent);
 
-                    ApplyDragToChildren(dragEvent);
+                    ApplyDragToChildren(xDelta, yDelta);
+                    this.PositionOffset = _previousPositionOffset;
```
Comment (Wouterdek) [inline NodeNetwork/Views/Controls/DragCanvas.cs:123]: Perhaps I'm missing something here, but isn't `ApplyDragToChildren` called automatically upon setting `PositionOffset`, making the function call redundant?

File: NodeNetwork/Views/Controls/DragCanvas.cs [hunk 7]
```diff
@@ -150,40 +168,100 @@ private void ApplyDragToChildren(DragMoveEventArgs drag)
...
+
+        public double MinZoomFactor
+        {
+            get { return (double)GetValue(MinZoomFactorProperty); }
+            set { SetValue(MinZoomFactorProperty, value); }
+        }
+
+        private static void MinZoomFactorChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if ((double)e.NewValue <= 0)
```
Comment (Wouterdek) [inline NodeNetwork/Views/Controls/DragCanvas.cs:254]: I like this precondition check, maybe add a NaN check too? Also, add these checks to MaxZoomFactor as well.

File: NodeNetwork/Views/Controls/DragCanvas.cs [hunk 8]
```diff
@@ -150,40 +168,100 @@ private void ApplyDragToChildren(DragMoveEventArgs drag)
...
+        {
+            var canvas = (DragCanvas)d;
+            var binding = BindingOperations.GetBindingExpression(canvas, ZoomFactorProperty);
+            binding?.UpdateTarget();
+        }
+        #endregion
+
+        #region MinZoomFactor
+        public static readonly DependencyProperty MinZoomFactorProperty = DependencyProperty.Register(nameof(MinZoomFactor),
+            typeof(double), typeof(DragCanvas), new FrameworkPropertyMetadata(1d, MinZoomFactorChanged));
```
Comment (Wouterdek) [inline NodeNetwork/Views/Controls/DragCanvas.cs:244]: The default values for `MinZoomFactor` and `MaxZoomFactor` need some tweaking. I think you used the values from `MinWheelOffset` and `MaxWheelOffset`, but wheel offset != zoom factor. With the current values, it is for example impossible to zoom out.

**Review by Wouterdek — CHANGES_REQUESTED** [review]
Hi, thank you for your contribution! I like what you've added, but I also have some notes that I've posted inline to the changelist.

File: NodeNetwork/ViewModels/NetworkViewModel.cs (hunk 1 above)
Comment (danielcaceresm) [reply NodeNetwork/ViewModels/NetworkViewModel.cs]: You're right, this is not necessary.

File: NodeNetwork/ViewModels/NetworkViewModel.cs (hunk 2 above)
Comment (danielcaceresm) [reply NodeNetwork/ViewModels/NetworkViewModel.cs]: I've fixed this.

File: NodeNetwork/ViewModels/NetworkViewModel.cs (hunk 3 above)
Comment (danielcaceresm) [reply NodeNetwork/ViewModels/NetworkViewModel.cs]: I've fixed this.

File: NodeNetwork/ViewModels/NetworkViewModel.cs (hunk 4 above)
Comment (danielcaceresm) [reply NodeNetwork/ViewModels/NetworkViewModel.cs]: I renamed the property to DragOffset as you suggested which is more descriptive.

File: NodeNetwork/Views/Controls/DragCanvas.cs (hunk 5 above)
Comment (danielcaceresm) [reply NodeNetwork/Views/Controls/DragCanvas.cs]: I renamed the property to DragOffset as you suggested which is more descriptive.

File: NodeNetwork/Views/Controls/DragCanvas.cs (hunk 6 above)
Comment (danielcaceresm) [reply NodeNetwork/Views/Controls/DragCanvas.cs]: You're right, this is not necessary.

File: NodeNetwork/Views/Controls/DragCanvas.cs (hunk 7 above)
Comment (danielcaceresm) [reply NodeNetwork/Views/Controls/DragCanvas.cs]: I've improved the precondition with NaN and Inifinite checks.

File: NodeNetwork/Views/Controls/DragCanvas.cs (hunk 8 above)
Comment (danielcaceresm) [reply NodeNetwork/Views/Controls/DragCanvas.cs]: You're right, I have updated the default values to match the previous behavior.

**Comment by Wouterdek** [comment]
Alright, looks good. There's a few more small things, but I'll fix those after merging. Thanks for your contribution!

**Pull request merged**